src/lightning/fabric/strategies/model_parallel.py

```diff
@@ -0,0 +1,331 @@
+            )
+        self._device_mesh = init_device_mesh(
+            device_type=self.root_device.type,
+            mesh_shape=(self._data_parallel_size, self._tensor_parallel_size),
+            mesh_dim_names=("data_parallel", "tensor_parallel"),
+        )
+
+    def _get_process_group_backend(self) -> str:
+        return self._process_group_backend or _get_default_process_group_backend_for_device(self.root_device)
+
+    def _set_world_ranks(self) -> None:
+        if self.cluster_environment is not None:
+            self.cluster_environment.set_global_rank(self.node_rank * self.num_processes + self.local_rank)
+            self.cluster_environment.set_world_size(self.num_nodes * self.num_processes)
+        # `LightningEnvironment.set_global_rank` will do this too, but we cannot rely on that implementation detail
+        # additionally, for some implementations, the setter is a no-op, so it's safer to access the getter
+        rank_zero_only.rank = utils_rank_zero_only.rank = self.global_rank
+
+
+def _materialize_module(module: Module, device: torch.device) -> None:
+    # Reference: https://github.com/pytorch/torchtitan/blob/main/docs/fsdp.md#meta-device-initialization
+    # TODO: Introduce `Fabric.materialize(module)` to give user control when materialization should happen
+    # TODO: Make `torchmetrics.Metric` compatible with the `to_empty()` + `reset_parameters()` semantics
+    if not _has_meta_device_parameters_or_buffers(module):
+        return
+
+    module.to_empty(device=device)  # has to be called on the root module
+
+    uninitialized_modules = set()
+    for submodule in module.modules():
+        if all(False for _ in itertools.chain(submodule.parameters(recurse=False), submodule.buffers(recurse=False))):
+            # module has no parameters or buffers
+            continue
+        if callable(reset_method := getattr(submodule, "reset_parameters", None)):
+            reset_method()
+        else:
+            uninitialized_modules.add(type(submodule).__name__)
+
+    if uninitialized_modules:
+        rank_zero_warn(
+            "Parameter initialization incomplete. The following modules have parameters or buffers with uninitialized"
+            " memory because they don't define a `reset_parameters()` method for re-initialization:"
+            f" {', '.join(uninitialized_modules)}"
+        )
+
+
+class _ParallelBackwardSyncControl(_BackwardSyncControl):
+    @override
+    def no_backward_sync(self, module: Module, enabled: bool) -> ContextManager:
+        """Blocks gradient synchronization inside the FSDP2 modules."""
+        return _FSDPNoSync(module=module, enabled=enabled)
+
+
+class _FSDPNoSync(ContextManager):
+    def __init__(self, module: Module, enabled: bool) -> None:
+        self._module = module
+        self._enabled = enabled
+
+    def _set_requires_grad_sync(self, requires_grad_sync: bool) -> None:
+        from torch.distributed._composable.fsdp import FSDP
```

Comment:
In the strategy readme, we should clarify that it is meant to be used with this FSDP implementation particularly and the tensor parallel implementation in PyTorch. Anything else might result in issues (like checkpointing, or no sync, etc)

Reply:
Done. I also added a check in setup_module that raises an error if FSDPv1 modules are found.